### ISSUE TITLE
allow attribute change according to role or method

### ISSUE THREAD

**Issue by nlhommet** [api-platform/api-platform#773]
Is it possible to allow attribute change according to role or method(POST, PUT)?
```
<?php
class User{

/**
* Allow for ROLE_USER in POST
* Not allow for ROLE_USER in PUT
* Allow for ROLE_ADMIN in POST and PUT
*/
private $username 

...
}
```

**Comment by dunglas** [comment]
You can use a serialization context builder to do that: https://api-platform.com/docs/core/serialization/#changing-the-serialization-context-dynamically

It would be nice to have built-in support for this, PR welcome (ping @vincentchalamon).

**Comment by Soullivaneuh** [comment]
As asked in https://github.com/api-platform/core/issues/2118#issuecomment-409874809:

> Will this method have any influence on the swagger/doc generation?

**Comment by fredericbarthelet** [comment]
Would love to help on that matter :)
I got to write quite a few context builder decorators on my projects and would benefit as well from using configuration to declare exposed attributes based on role.

Before starting, I would like your input on the following design consideration

If i understood well, current design pattern for context builder is decorator. Core currently contains 2 context builders. The originator src/Serializer/SerializerContextBuilder.php and a first decorator src/Serializer/SerializerFilterContextBuilder.php. Implementing another layer of logic to build context[‘groups’] could be done with another service decorating the FilterContextBuilder. You would then have 3 layers as part of core bundle and still the possibility for the developper to decorate further more the context builder.

Another solution would be to change ContextBuilderInterface to require implementation of methods `enrichContext` and `supportRequest`. Service SerializerContextBuilder.php would not implement this interface but would instead iterate over other services implementing the interface and pass on previously modified context array to subsequent context builders. This change would break API contract and force developers to change the implementation of their current context builder but would get closer to existing implementation of SerializerInterface, DataTransformerInterface, …

@dunglas what do you think of the second solution ?

**Comment by teohhanhui** [comment]
@dunglas Based on our previous discussion, we would like to decouple `SerializerContextBuilderInterface` from the `Request`, right?

**Comment by alanpoulain** [comment]
Merged here: https://github.com/api-platform/core/pull/3503.
